src/common/hashtable/src/table0.rs

```diff
@@ -0,0 +1,377 @@
+        assume(!K::equals_zero(key));
+        let index = (hash as usize) & (self.entries.len() - 1);
+        for i in (index..self.entries.len()).chain(0..index) {
+            assume(i < self.entries.len());
+            if self.entries[i].is_zero() {
+                return None;
+            }
+            if self.entries[i].key.assume_init_ref().borrow() == key {
+                return Some(&mut self.entries[i]);
+            }
+        }
+        None
+    }
+    /// # Safety
+    ///
+    /// `key` doesn't equal to zero.
+    /// The resulted `MaybeUninit` should be initialized immedidately.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the hash table overflows.
+    #[inline(always)]
+    pub unsafe fn insert(&mut self, key: K) -> Result<&mut Entry<K, V>, &mut Entry<K, V>> {
+        self.insert_with_hash(key, key.hash())
+    }
+    /// # Safety
+    ///
+    /// `key` doesn't equal to zero.
+    /// The resulted `MaybeUninit` should be initialized immedidately.
+    /// Provided hash is correct.
+    ///
+    /// # Panics
+    /// The hashtable is full.
+    #[inline(always)]
+    pub unsafe fn insert_with_hash(
+        &mut self,
+        key: K,
+        hash: u64,
+    ) -> Result<&mut Entry<K, V>, &mut Entry<K, V>> {
+        assume(!K::equals_zero(&key));
+        let index = (hash as usize) & (self.entries.len() - 1);
+        for i in (index..self.entries.len()).chain(0..index) {
+            assume(i < self.entries.len());
+            if self.entries[i].is_zero() {
+                self.len += 1;
+                self.entries[i].key.write(key);
+                return Ok(&mut self.entries[i]);
+            }
+            if self.entries[i].key.assume_init_ref() == &key {
+                return Err(&mut self.entries[i]);
+            }
+        }
+        panic!("the hash table overflows")
+    }
+    pub fn iter(&self) -> Table0Iter<'_, K, V> {
+        Table0Iter {
+            slice: self.entries.as_ref(),
+            i: 0,
+        }
+    }
+    pub fn iter_mut(&mut self) -> Table0IterMut<'_, K, V> {
+        Table0IterMut {
+            slice: self.entries.as_mut(),
+            i: 0,
+        }
+    }
+    pub fn grow(&mut self, shift: u8) {
+        let old_capacity = self.entries.len();
+        let new_capacity = self.entries.len() << shift;
+        unsafe {
+            self.entries.grow_zeroed(new_capacity);
+        }
+        for i in 0..old_capacity {
+            unsafe {
+                assume(i < self.entries.len());
+            }
+            if K::is_zero(&self.entries[i].key) {
+                continue;
+            }
+            let key = unsafe { self.entries[i].key.assume_init_ref() };
+            let hash = K::hash(key);
+            let index = (hash as usize) & (self.entries.len() - 1);
+            for j in (index..self.entries.len()).chain(0..index) {
+                unsafe {
+                    assume(j < self.entries.len());
```

Comment:
what's this assume used for?

Reply:
To skip bound checking in the following lines.